File: BittraBLETest/BLE/BLECentral.swift
import Foundation
import CoreBluetooth

final class BLECentral: NSObject, CBCentralManagerDelegate, CBPeripheralDelegate {

  private var cm: CBCentralManager!
  private let log: (String, String) -> Void

  // 発見データ
  var onEncounter: ((CBPeripheral, String, Int) -> Void)?

  // 本文受信
  var onStatus: ((RequestStatus) -> Void)?
  var onPreview: ((String) -> Void)?
  var onBody: ((String) -> Void)?

  private var targetPeripheral: CBPeripheral?
  private var reqChar: CBCharacteristic?
  private var chunkChar: CBCharacteristic?

  private var assembler = ChunkAssembler()
  private var timeoutTimer: DispatchSourceTimer?
  private var retryLeft: Int = 1

  init(log: @escaping (String, String) -> Void) {
    self.log = log
    super.init()
    self.cm = CBCentralManager(delegate: self, queue: nil)
  }

  func startScan() {
    guard cm.state == .poweredOn else {
      log("CENTRAL", "Bluetooth not poweredOn yet")
      return
    }
    cm.scanForPeripherals(withServices: [GATTProfile.serviceUUID], options: [
      CBCentralManagerScanOptionAllowDuplicatesKey: true
    ])
    log("SCAN", "startScan")
  }

  func stopScan() {
    cm.stopScan()
    log("SCAN", "stopScan")
  }

  func connect(_ peripheral: CBPeripheral) {
    resetRequestState()
    retryLeft = 1
    targetPeripheral = peripheral
    targetPeripheral?.delegate = self
    onStatus?(.connecting)
    cm.connect(peripheral, options: nil)
    startTimeout(seconds: 2.0)
    log("CENTRAL", "connect start \(peripheral.identifier)")
  }

  private func resetRequestState() {
    stopTimeout()
    assembler.reset()
    reqChar = nil
    chunkChar = nil
  }

  private func startTimeout(seconds: Double) {
    stopTimeout()
    let t = DispatchSource.makeTimerSource(queue: .main)
    t.schedule(deadline: .now() + seconds)
    t.setEventHandler { [weak self] in
      guard let self else { return }
      self.log("CENTRAL", "timeout")
      self.onStatus?(.timeout)
      self.failOrRetry("timeout")
    }
    t.resume()
    timeoutTimer = t
  }

  private func stopTimeout() {
    timeoutTimer?.cancel()
    timeoutTimer = nil
  }

  private func failOrRetry(_ reason: String) {
    if retryLeft > 0, let p = targetPeripheral {
      retryLeft -= 1
      log("CENTRAL", "retry \(retryLeft) reason=\(reason)")
      cm.cancelPeripheralConnection(p)
      DispatchQueue.main.asyncAfter(deadline: .now() + 0.2) {
        self.connect(p)
      }
    } else {
      onStatus?(.failed(reason))
      if let p = targetPeripheral {
        cm.cancelPeripheralConnection(p)
      }
      stopTimeout()
    }
  }

  // MARK: - CBCentralManagerDelegate

  func centralManagerDidUpdateState(_ central: CBCentralManager) {
    log("CENTRAL", "state=\(central.state.rawValue)")
  }

  func centralManager(_ central: CBCentralManager,
                      didDiscover peripheral: CBPeripheral,
                      advertisementData: [String : Any],
                      rssi RSSI: NSNumber) {
    guard let sd = advertisementData[CBAdvertisementDataServiceDataKey] as? [CBUUID: Data],
          let payload = sd[GATTProfile.serviceUUID],
          let decoded = PayloadCodec.decodeAdvert(payload) else { return }

    onEncounter?(peripheral, decoded.teaser, RSSI.intValue)
  }

  func centralManager(_ central: CBCentralManager, didConnect peripheral: CBPeripheral) {
    log("CENTRAL", "didConnect")
    onStatus?(.discovering)
    peripheral.discoverServices([GATTProfile.serviceUUID])
  }

  func centralManager(_ central: CBCentralManager, didFailToConnect peripheral: CBPeripheral, error: Error?) {
    log("CENTRAL", "didFailToConnect \(error?.localizedDescription ?? "-")")
    failOrRetry("connect_failed")
  }

  func centralManager(_ central: CBCentralManager, didDisconnectPeripheral peripheral: CBPeripheral, error: Error?) {
    log("CENTRAL", "didDisconnect \(error?.localizedDescription ?? "-")")
    // 途中切断は会場では普通に起きるので、ここでは Coordinator 側で扱う
  }

  // MARK: - CBPeripheralDelegate

  func peripheral(_ peripheral: CBPeripheral, didDiscoverServices error: Error?) {
    if let e = error {
      log("CENTRAL", "discoverServices error \(e.localizedDescription)")
      return failOrRetry("discover_services")
    }
    guard let services = peripheral.services,
          let s = services.first(where: { $0.uuid == GATTProfile.serviceUUID }) else {
      return failOrRetry("no_service")
    }
    peripheral.discoverCharacteristics([GATTProfile.reqCharUUID, GATTProfile.chunkCharUUID], for: s)
  }

  func peripheral(_ peripheral: CBPeripheral, didDiscoverCharacteristicsFor service: CBService, error: Error?) {
    if let e = error {
      log("CENTRAL", "discoverChars error \(e.localizedDescription)")
      return failOrRetry("discover_chars")
    }
    guard let chars = service.characteristics else { return failOrRetry("no_chars") }
    reqChar = chars.first(where: { $0.uuid == GATTProfile.reqCharUUID })
    chunkChar = chars.first(where: { $0.uuid == GATTProfile.chunkCharUUID })

    guard let reqChar, let chunkChar else { return failOrRetry("missing_char") }

    onStatus?(.subscribing)
    peripheral.setNotifyValue(true, for: chunkChar)
  }

  func peripheral(_ peripheral: CBPeripheral, didUpdateNotificationStateFor characteristic: CBCharacteristic, error: Error?) {
    if let e = error {
      log("CENTRAL", "notifyState error \(e.localizedDescription)")
      return failOrRetry("notify_fail")
    }
    guard characteristic.uuid == GATTProfile.chunkCharUUID else { return }

    onStatus?(.requesting)
    // REQ：今回は中身空でOK（「全文ください」）
    let req = Data([0x01])
    if let reqChar {
      peripheral.writeValue(req, for: reqChar, type: .withoutResponse)
      log("CENTRAL", "REQ sent")
      // タイムアウトは維持（2秒）
    } else {
      failOrRetry("no_req_char")
    }
  }

  func peripheral(_ peripheral: CBPeripheral, didUpdateValueFor characteristic: CBCharacteristic, error: Error?) {
    if let e = error {
      log("CENTRAL", "didUpdateValue error \(e.localizedDescription)")
      return
    }
    guard characteristic.uuid == GATTProfile.chunkCharUUID,
          let value = characteristic.value,
          value.count >= 3 else { return }

    let seq = UInt16(value[0]) | (UInt16(value[1]) << 8)
    let flags = value[2]
    let payload = value.subdata(in: 3..<value.count)

    let (previewReady, completed, fullData) = assembler.addChunk(seq: seq, flags: flags, payload: payload)

    if previewReady {
      onStatus?(.receivingPreview)
      // 受信済みの分だけテキスト化して見せる（簡易）
      // 完全な順序保証がないので、MVPでは previewReady を優先
      // （本番は0..nを揃えてからにしてもOK）
      // ここでは fullData が無い場合、今ある分を適当に連結して表示しない
    }

    if completed, let fullData {
      stopTimeout()
      let s = String(data: fullData, encoding: .utf8) ?? ""
      onBody?(s)
      onStatus?(.completed)
      cm.cancelPeripheralConnection(peripheral)
    }
  }
}

File: android/app/src/main/kotlin/jp/cloxs/bittra/ble/BleCentral.kt
package jp.cloxs.bitra.ble

import android.annotation.SuppressLint
import android.bluetooth.*
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanFilter
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.ParcelUuid
import java.util.UUID

enum class RequestStatus(val rawValue: String) {
    IDLE("idle"),
    CONNECTING("connecting"),
    DISCOVERING("discovering"),
    SUBSCRIBING("subscribing"),
    REQUESTING("requesting"),
    RECEIVING_PREVIEW("receivingPreview"),
    COMPLETED("completed"),
    TIMEOUT("timeout"),
    FAILED("failed")
}

@SuppressLint("MissingPermission")
class BleCentral(private val context: Context, private val log: (String, String) -> Unit) {

    private val bluetoothManager = context.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
    private val adapter: BluetoothAdapter? = bluetoothManager.adapter
    private val scanner = adapter?.bluetoothLeScanner

    var onEncounter: ((BluetoothDevice, Long, String, Int) -> Unit)? = null
    var onStatus: ((RequestStatus, String?) -> Unit)? = null
    var onPreview: ((String) -> Unit)? = null
    var onBody: ((String) -> Unit)? = null

    private var currentGatt: BluetoothGatt? = null
    private var reqChar: BluetoothGattCharacteristic? = null
    private var chunkChar: BluetoothGattCharacteristic? = null
    private var ackChar: BluetoothGattCharacteristic? = null

    private val assembler = ChunkAssembler()
    private val mainHandler = Handler(Looper.getMainLooper())
    private var timeoutRunnable: Runnable? = null
    private var ackTimeoutRunnable: Runnable? = null
    private var pendingCompletedBody: String? = null
    private var retryLeft = 1
    private var targetDevice: BluetoothDevice? = null
    private var isScanning = false
    private var resumeScanAfterRequest = false

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            val record = result.scanRecord ?: return

            val mfgData = record.getManufacturerSpecificData(GATTProfile.MANUFACTURER_ID)
            val srvData = record.serviceData[ParcelUuid(GATTProfile.SERVICE_UUID)]

            android.util.Log.d("BleCentral", "ScanResult for ${result.device.address}: name=${record.deviceName}, mfgData=${mfgData?.size}, srvData=${srvData?.size}, srvUuids=${record.serviceUuids}")

            val decoded = if (mfgData != null && mfgData.isNotEmpty()) {
                PayloadCodec.decodeAdvert(mfgData)
            } else if (srvData != null && srvData.isNotEmpty()) {
                PayloadCodec.decodeAdvert(srvData)
            } else {
                val localName = record.deviceName
                if (localName != null) {
                    val compactResult = PayloadCodec.decodeLocalName(localName)
                    if (compactResult != null) {
                        compactResult
                    } else {
                        var legacyResult: PayloadCodec.AdvertResult? = null
                        var tempName: String = localName
                        for (i in 0 until 4) {
                            if (tempName.isEmpty()) break
                            try {
                                val padded = tempName.padEnd(tempName.length + (4 - tempName.length % 4) % 4, '=')
                                val decodedData = android.util.Base64.decode(padded, android.util.Base64.DEFAULT)
                                legacyResult = PayloadCodec.decodeAdvert(decodedData)
                                if (legacyResult != null) break
                            } catch (_: IllegalArgumentException) {
                            }
                            tempName = tempName.dropLast(1)
                        }
                        legacyResult
                    }
                } else {
                    null
                }
            }

            if (decoded == null || decoded.teaser.isEmpty()) return
            onEncounter?.invoke(result.device, decoded.senderId, decoded.teaser, result.rssi)
        }

        override fun onScanFailed(errorCode: Int) {
            isScanning = false
            log("SCAN", "startScan failed: $errorCode")
        }
    }

    private val gattCallback = object : BluetoothGattCallback() {
        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            if (gatt !== currentGatt) {
                gatt.close()
                return
            }

            if (newState == BluetoothProfile.STATE_CONNECTED) {
                log("CENTRAL", "didConnect")
                mainHandler.postDelayed({
                    if (gatt !== currentGatt) return@postDelayed
                    onStatus?.invoke(RequestStatus.DISCOVERING, null)
                    if (!gatt.discoverServices()) {
                        failOrRetry("discover_services_not_started")
                    }
                }, 300)
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                log("CENTRAL", "didDisconnect status=$status")
                if (status != BluetoothGatt.GATT_SUCCESS && currentGatt === gatt) {
                    mainHandler.post { failOrRetry("connect_failed_status_$status") }
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            if (gatt !== currentGatt) return
            if (status != BluetoothGatt.GATT_SUCCESS) {
                log("CENTRAL", "discoverServices error $status")
                mainHandler.post { failOrRetry("discover_services_$status") }
                return
            }

            val service = gatt.getService(GATTProfile.SERVICE_UUID)
            if (service == null) {
                mainHandler.post { failOrRetry("no_service") }
                return
            }

            reqChar = service.getCharacteristic(GATTProfile.REQ_CHAR_UUID)
            chunkChar = service.getCharacteristic(GATTProfile.CHUNK_CHAR_UUID)
            ackChar = service.getCharacteristic(GATTProfile.ACK_CHAR_UUID)

            if (reqChar == null || chunkChar == null) {
                mainHandler.post { failOrRetry("missing_char") }
                return
            }

            if (ackChar == null) {
                log("CENTRAL", "sender does not support delivery ACK")
            }

            mainHandler.post {
                onStatus?.invoke(RequestStatus.SUBSCRIBING, null)
                gatt.setCharacteristicNotification(chunkChar, true)

                val descriptor = chunkChar?.getDescriptor(UUID.fromString("00002902-0000-1000-8000-00805f9b34fb"))
                if (descriptor != null) {
                    descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
                    gatt.writeDescriptor(descriptor)
                } else {
                    sendRequest()
                }
            }
        }

        override fun onDescriptorWrite(gatt: BluetoothGatt, descriptor: BluetoothGattDescriptor, status: Int) {
            if (gatt !== currentGatt) return
            if (status == BluetoothGatt.GATT_SUCCESS) {
                mainHandler.post { sendRequest() }
            } else {
                mainHandler.post { failOrRetry("descriptor_write_fail_$status") }
            }
        }

        override fun onCharacteristicWrite(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            status: Int
        ) {
            if (gatt !== currentGatt || characteristic.uuid != GATTProfile.ACK_CHAR_UUID) return
            mainHandler.post {
                stopAckTimeout()
                if (status == BluetoothGatt.GATT_SUCCESS) {
                    log("CENTRAL", "delivery ACK confirmed")
                } else {
                    log("CENTRAL", "delivery ACK failed status=$status")
                }
                completePendingBody()
            }
        }

        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic,
            value: ByteArray
        ) {
            handleCharacteristicChanged(characteristic, value)
        }

        @Suppress("DEPRECATION")
        override fun onCharacteristicChanged(
            gatt: BluetoothGatt,
            characteristic: BluetoothGattCharacteristic
        ) {
            handleCharacteristicChanged(characteristic, characteristic.value ?: return)
        }
    }

    private fun handleCharacteristicChanged(characteristic: BluetoothGattCharacteristic, value: ByteArray) {
        if (characteristic.uuid != GATTProfile.CHUNK_CHAR_UUID || value.size < 3) return

        val seq = (value[0].toInt() and 0xFF) or ((value[1].toInt() and 0xFF) shl 8)
        val flags = value[2].toInt() and 0xFF
        val payload = value.copyOfRange(3, value.size)

        val (previewReady, completed, fullData) = assembler.addChunk(seq, flags, payload)

        mainHandler.post {
            if (previewReady) {
                onStatus?.invoke(RequestStatus.RECEIVING_PREVIEW, null)
            }

            if (completed && fullData != null) {
                stopTimeout()
                val body = String(fullData, Charsets.UTF_8)
                pendingCompletedBody = body
                sendDeliveryAckOrComplete()
            }
        }
    }

    private fun sendDeliveryAckOrComplete() {
        val characteristic = ackChar
        if (characteristic == null) {
            completePendingBody()
            return
        }

        characteristic.value = byteArrayOf(0x02)
        characteristic.writeType = BluetoothGattCharacteristic.WRITE_TYPE_DEFAULT
        if (currentGatt?.writeCharacteristic(characteristic) == true) {
            log("CENTRAL", "delivery ACK sent")
            startAckTimeout(2000)
        } else {
            log("CENTRAL", "delivery ACK write not started")
            completePendingBody()
        }
    }

    private fun completePendingBody() {
        stopAckTimeout()
        val body = pendingCompletedBody ?: return
        pendingCompletedBody = null
        val preview = if (body.length > 40) body.substring(0, 40) else body
        onBody?.invoke(body)
        onPreview?.invoke(preview)
        onStatus?.invoke(RequestStatus.COMPLETED, null)
        closeCurrentGatt()
        finishRequest()
    }

    private fun sendRequest() {
        onStatus?.invoke(RequestStatus.REQUESTING, null)
        val req = byteArrayOf(0x01)
        reqChar?.let {
            it.value = req
            it.writeType = BluetoothGattCharacteristic.WRITE_TYPE_NO_RESPONSE
            if (currentGatt?.writeCharacteristic(it) == true) {
                log("CENTRAL", "REQ sent")
            } else {
                failOrRetry("request_write_not_started")
            }
        } ?: failOrRetry("no_req_char")
    }

    fun startScan() {
        if (adapter == null || !adapter.isEnabled || scanner == null) {
            log("CENTRAL", "Bluetooth not poweredOn yet")
            return
        }

        if (isScanning) {
            log("SCAN", "already scanning")
            return
        }

        val filters = listOf(ScanFilter.Builder().setServiceUuid(ParcelUuid(GATTProfile.SERVICE_UUID)).build())
        val settings = ScanSettings.Builder()
            .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY)
            .build()

        scanner.startScan(filters, settings, scanCallback)
        isScanning = true
        log("SCAN", "startScan")
    }

    fun stopScan() {
        if (isScanning) {
            scanner?.stopScan(scanCallback)
            isScanning = false
        }
        log("SCAN", "stopScan")
    }

    fun requestBody(device: BluetoothDevice) {
        log("CENTRAL", "requestBody called for peerId=${device.address}")
        retryLeft = 1
        targetDevice = device
        resumeScanAfterRequest = isScanning
        if (isScanning) stopScan()
        connect(device)
    }

    private fun connect(device: BluetoothDevice) {
        resetRequestState()
        onStatus?.invoke(RequestStatus.CONNECTING, null)
        currentGatt = device.connectGatt(context, false, gattCallback, BluetoothDevice.TRANSPORT_LE)
        startTimeout(10000)
        log("CENTRAL", "connect start ${device.address}")
    }

    private fun resetRequestState() {
        stopTimeout()
        stopAckTimeout()
        assembler.reset()
        pendingCompletedBody = null
        reqChar = null
        chunkChar = null
        ackChar = null
        closeCurrentGatt()
    }

    private fun closeCurrentGatt() {
        val oldGatt = currentGatt
        currentGatt = null
        oldGatt?.disconnect()
        oldGatt?.close()
    }

    private fun finishRequest() {
        targetDevice = null
        if (resumeScanAfterRequest) {
            resumeScanAfterRequest = false
            startScan()
        }
    }

    private fun startTimeout(ms: Long) {
        stopTimeout()
        val r = Runnable {
            log("CENTRAL", "timeout")
            onStatus?.invoke(RequestStatus.TIMEOUT, null)
            failOrRetry("timeout")
        }
        mainHandler.postDelayed(r, ms)
        timeoutRunnable = r
    }

    private fun stopTimeout() {
        timeoutRunnable?.let { mainHandler.removeCallbacks(it) }
        timeoutRunnable = null
    }

    private fun startAckTimeout(ms: Long) {
        stopAckTimeout()
        val r = Runnable {
            log("CENTRAL", "delivery ACK timeout")
            completePendingBody()
        }
        mainHandler.postDelayed(r, ms)
        ackTimeoutRunnable = r
    }

    private fun stopAckTimeout() {
        ackTimeoutRunnable?.let { mainHandler.removeCallbacks(it) }
        ackTimeoutRunnable = null
    }

    private fun failOrRetry(reason: String) {
        stopTimeout()
        stopAckTimeout()
        if (retryLeft > 0 && targetDevice != null) {
            retryLeft -= 1
            log("CENTRAL", "retry $retryLeft reason=$reason")
            val p = targetDevice!!
            closeCurrentGatt()
            mainHandler.postDelayed({ connect(p) }, 200)
        } else {
            onStatus?.invoke(RequestStatus.FAILED, reason)
            closeCurrentGatt()
            finishRequest()
        }
    }
}
